### PR TITLE
planner: fix the panic in joinReOrderSolver.optimizeRecursive

### LINKED ISSUE #23030: core.(*joinReOrderSolver).optimizeRecursive panic with the runtime error "index out of range [6] with length 6"
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
download the [sqls](https://github.com/pingcap/tidb/files/6066079/1.txt)
```sql
create database llh;
source 1.txt;
SELECT count(1) FROM view_10 AS tmp1 JOIN view_8 AS tmp2  WHERE (tmp2.col_1);
```

### 2. What did you expect to see? (Required)
successfully
### 3. What did you see instead (Required)
ERROR 1105 (HY000): runtime error: index out of range [4] with length 4
The tidb log is
```
[err="runtime error: index out of range [6] with length 6"] [stack="goroutine 7765713 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3f974a0, 0xc0031619e0, 0xc0013bad00)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:731 +0xee
panic(0x373c600, 0xc002cd4ac0)
\t/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/pingcap/tidb/planner/core.(*joinReOrderSolver).optimizeRecursive(0x5eea3e0, 0x3fef940, 0xc0013bae00, 0x3feae40, 0xc000aad180, 0x0, 0x0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_join_reorder.go:95 +0xa4d
github.com/pingcap/tidb/planner/core.(*joinReOrderSolver).optimizeRecursive(0x5eea3e0, 0x3fef940, 0xc0013bae00, 0x3feab40, 0xc0023d5e00, 0x0, 0x0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_join_reorder.go:112 +0x894
github.com/pingcap/tidb/planner/core.(*joinReOrderSolver).optimizeRecursive(0x5eea3e0, 0x3fef940, 0xc0013bae00, 0x3feb440, 0xc001a7c000, 0xc001a7c000, 0xc000e48b70, 0x36639a0, 0xc000e48b70)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_join_reorder.go:112 +0x894
github.com/pingcap/tidb/planner/core.(*joinReOrderSolver).optimize(0x5eea3e0, 0x3f974a0, 0xc001020d50, 0x3feb440, 0xc001a7c000, 0x3feb440, 0xc001a7c000, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_join_reorder.go:58 +0x5a
github.com/pingcap/tidb/planner/core.logicalOptimize(0x3f974a0, 0xc001020d50, 0x3ff6, 0x3feb440, 0xc001a7c000, 0x603db60b, 0xc00185a450, 0x13c37c6, 0x603db60b)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/optimizer.go:194 +0x116
github.com/pingcap/tidb/planner/core.DoOptimize(0x3f974a0, 0xc001020d50, 0x3fef940, 0xc0013bae00, 0x1ff6, 0x3feb440, 0xc001a7c000, 0x0, 0x0, 0x0, ...)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/optimizer.go:128 +0x77
github.com/pingcap/tidb/planner.optimize(0x3f974a0, 0xc001020d50, 0x3fef940, 0xc0013bae00, 0x3fb6020, 0xc0013bb700, 0x3fe1b40, 0xc0019f1130, 0x0, 0x0, ...)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/optimize.go:275 +0x595
github.com/pingcap/tidb/planner.Optimize(0x3f974a0, 0xc001020d50, 0x3fef940, 0xc0013bae00, 0x3fb6020, 0xc0013bb700, 0x3fe1b40, 0xc0019f1130, 0x0, 0x0, ...)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/optimize.go:123 +0x491
github.com/pingcap/tidb/executor.(*Compiler).Compile(0xc00185ad78, 0x3f974a0, 0xc001020d50, 0x3fbdee0, 0xc0013bb700, 0x0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/compiler.go:62 +0x294
github.com/pingcap/tidb/session.(*session).ExecuteStmt(0xc0013bae00, 0x3f974a0, 0xc0010209f0, 0x3fbdee0, 0xc0013bb700, 0x0, 0x0, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/session/session.go:1540 +0x2c4
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt(0xc003161fb0, 0x3f974a0, 0xc0010209f0, 0x3fbdee0, 0xc0013bb700, 0xc0021a7fb0, 0x3f974a0, 0xc0010209f0, 0xc00170d148)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/driver_tidb.go:212 +0x68
github.com/pingcap/tidb/server.(*clientConn).handleStmt(0xc0013bad00, 0x3f974a0, 0xc0010209f0, 0x3fbdee0, 0xc0013bb700, 0x5eeaab0, 0x0, 0x0, 0x1, 0x0, ...)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1607 +0x134
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc0013bad00, 0x3f973e0, 0xc001b178c0, 0xc002437a91, 0x4c, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1481 +0x44f
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc0013bad00, 0x3f973e0, 0xc001b178c0, 0xc002437a90, 0x4d, 0x4c, 0x0, 0x0)
\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1027 +0x698
github.com/pingcap/tidb/server.(*clientConn).Run(0xc0013bad00, 0x3f974a0, 0xc0031619e0)
\t/home/jenkins/agent/workspace/tidb_ma"]
```
### 4. What is your TiDB version? (Required)
Release Version: v4.0.0-beta.2-2215-g6bd6df7d1
Edition: Community
Git Commit Hash: 6bd6df7d1ec0b0a2056a5d40aca8261c5cbfa871
Git Branch: master
UTC Build Time: 2021-03-01 08:36:48
GoVersion: go1.15.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: close #23030 

Problem Summary:
When we want to compare the schema with the origin schema, we forget to compare the length.

### What is changed and how it works?

When we want to compare the schema,  we check the length first.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- fix the panic in joinReOrderSolver.optimizeRecursive

### CODE REVIEW

**Comment by Reminiscent** [comment]
@qw4990 @winoros PTAL

File: planner/core/rule_join_reorder.go [hunk 1]
```diff
@@ -91,10 +91,15 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
 			return nil, err
 		}
 		schemaChanged := false
-		for i, col := range p.Schema().Columns {
-			if !col.Equal(nil, originalSchema.Columns[i]) {
-				schemaChanged = true
-				break
+		// We compare the length of the column in the schema first.
```
Comment (winoros) [inline planner/core/rule_join_reorder.go:94]: Could you explain with the added case to show which column is added in the final schema?

Comment (zimulala) [reply planner/core/rule_join_reorder.go:94]: PTAL @Reminiscent

**Comment by Reminiscent** [comment]
After investigation, it was found that this bug has been fixed on `master` and `release-5.0`.

**Pull request closed without merging**